### ISSUE TITLE
Worldmap - not possible to use multiple worldmaps with their own source-files

### ISSUE THREAD

**Issue by kafos** [NagVis/nagvis#139]
it is not possible to create a second worldmap with a different .db file. The next created map has the same hosts like first one. It should be possible to create worldmaps with different sources

**Comment by LarsMichelsen** [comment]
The idea behind the worldmap was that each object has always only a single position on the planet and will not be on another position at the same time. That should also be true for you, right?

What do you want to archieve with such a change?

**Comment by Dennis14e** [comment]
In our company we use the worldmap to map our network connections.
Multiple maps (or multiple layers) would make sense, e.g. our DSL and directional radio antennas (dt. Richtfunkantennen) could be separated (= looks clearer).

**Comment by kafos** [comment]
@LarsMichelsen in our company we have different kind/category of host, which should be displayed in different worldmaps. And this is not possible now because everything is on one map. It would be really nice if you can implement the possibility to choose a config/db-file where the map is storing the data.

**Comment by angelfoga** [comment]
Yes, I am in the same situation. 
My company needs to have one map per customer, we are monitoring different stations and we want to give every one a personal map where they can see the status of their services/hosts with one look. 
We are currently using Nagios 4.4.1 and out of ideas, I was very excited at first with this tool but for us it's just useless without the user-view map asigned to the hostsgroup that he owns.
If someone is in the same situation or have any ideas, please let me know.

**Comment by T100D** [comment]
@angelfoga and others.

To serve multiple customers one could install multiple instances of Nagvis all in their own directory.

**Comment by kafos** [comment]
@T100D that is maybe just a workaround. It's just not making sense beeing able to create multiple worldmaps in one instance but not beeing able to use different config files for each map

**Comment by angelfoga** [comment]
@kafos I agree, but it's a **free** tool and personaly I don't know how hard would be to implement this. 
@T100D Thanks! I'll try this workaround, could be the solution for me.

**Comment by rcavezzale** [comment]
@LarsMichelsen 
Sorry, do you think that you will realize this enhancement?

**Comment by ekrichbaum** [comment]
If you add the objects to the map file rather than the db, they appear only on that map.  Not as editable/dynamic but would allow you to have an object (like a datacenter) appear on every map but a different object only on specific maps.

**Comment by neperpbl3** [comment]
Hello!

Thank you for advice!

For use multiple worldmaps was modified PHP files GlobalMapCfg.php, worldmap.php.

The database name 'worldmap.db' has been changed to "worldmap-$mapnamedb.db".



```
/usr/local/nagvis/share/server/core/classes# diff -u GlobalMapCfg-ORIGINAL GlobalMapCfg.php
--- GlobalMapCfg-ORIG   2021-10-12 07:23:04.000000000 +0700
+++ GlobalMapCfg.php    2021-11-05 00:33:48.867000221 +0700
@@ -59,6 +59,9 @@
         $this->name = $name;
         $this->isView = $isView;

+// Modifited for use multiple worldmaps with their own source DB files
+       global $mapnamedb;
+       $mapnamedb = $name;
         if(self::$validConfig == null)
             $this->fetchValidConfig();
```




```
/usr/local/nagvis/share/server/core/sources# diff -u worldmap-ORIGINAL worldmap.php
--- worldmap-ORIG       2021-11-03 11:47:16.000000000 +0700
+++ worldmap.php        2021-11-05 00:34:31.035447522 +0700
@@ -163,11 +163,13 @@
 }

 function worldmap_init_db() {
-    global $DB;
+    global $DB, $mapnamedb;
     if ($DB !== null)
         return; // only init once
     $DB = new CorePDOHandler();
-    if (!$DB->open('sqlite', array('filename' => cfg('paths', 'cfg').'worldmap.db'), null, null))
+    // Modifited for use multiple worldmaps with their own source DB files
+//    if (!$DB->open('sqlite', array('filename' => cfg('paths', 'cfg').'worldmap.db'), null, null))
+    if (!$DB->open('sqlite', array('filename' => cfg('paths', 'cfg')."worldmap-$mapnamedb.db"), null, null))
         throw new NagVisException(l('Unable to open worldmap database ([DB]): [MSG]',
             Array('DB' => $DB->getDSN(),
                   'MSG' => json_encode($DB->error()))));
@@ -456,7 +458,11 @@
 }

 function changed_worldmap($MAPCFG, $compare_time) {
-    $db_path = cfg('paths', 'cfg').'worldmap.db';
+// Modifited for use multiple worldmaps with their own source DB files
+//    $db_path = cfg('paths', 'cfg').'worldmap.db';
+    global $mapnamedb;
+    $db_path = cfg('paths', 'cfg')."worldmap-$mapnamedb.db";
+
     return !file_exists($db_path) || filemtime($db_path) > $compare_time;
 }
```

**Comment by ekrichbaum** [comment]
That's actually pretty slick. If you don't use the db have have a cfg file you could probably have your objects unique as well, but this is more elegant.

**Comment by LarsMichelsen** [comment]
The approach is hacky, but you may want to clean it up a bit and send a push request. I'd be happy to add it. One suggestion would be to hand over $map_name from the call sites of worldmap_init_db() as argument instead of using the global variable $mapnamedb.

A more minimalistic approach would be to add a mode where the worldmap coordinates are read from the map configuration as normal attributes, but that would require more changes.

**Comment by neperpbl3** [comment]
> . I'd be happy to add it. 

I'm not a programmer. I did everything I could. Please add the correct way. This is a very necessary extension. 

Please!!!
